Graphs/js/main-directed.js

```diff
@@ -827,3 +827,143 @@ function myPageRank() {
         animateNodes(ranks, 1000);
     }
 }
+
+function heldKarp() {
+    stopAnimation();
+    let root;
+    const [wm, id_index] = getWM(cy, false, true);
+
+    if(!isComplete()){
+        alert('Graph not complete. Please set the probability that two nodes are connected to 1.');
+        return;
+    } 
+
+    if (cy.elements(':selected').length >= 1) {
+        root = cy.elements(':selected')[0];
+    } else {
+        root = prompt('Please enter the id of the start node.\n', '1');
+        root = parseInt(root) - 1;
+    }
+
+    clearCyStyle();
+    
+    const nodes = getAllNodes(cy);
+
+    cy.elements(':grabbable').unselect();
+
+    heldKarpHelp(wm, root);
+}
+
+/**
+ *
+ * @param {number[][]} wm weight matrix
+ * @param {number} x starting point
+ */
+function heldKarpHelp(wm, x) {
+    let n = wm.length;
+    var memo = new Map();
+
+    let nums = new Array(n);
+    for (let i = 0; i < n; i++) {
+        nums[i] = i;
+    }
+    nums.splice(x, 1);
+
+    let rst = new Array(n - 1);
+    let min = Infinity;
+    let minArr;
+
+    for (let i = 0; i < nums.length; i++) {
+        let tempNums = nums.concat();
+        tempNums.splice(i, 1);
+        let h = heldKarpPath(nums[i], x, tempNums, wm, memo);
+        rst[i] = wm[x][nums[i]] + h.distance;
+
+        if (wm[x][nums[i]] + h.distance < min) {
+            min = wm[x][nums[i]] + h.distance;
+            minArr = h.arr.concat();
+        }
+    }
+
+    let minArrInv = [];
```

Comment:
There is actually a reverse method on arrays.
Note that .reverse() is done in-place, so you need to copy the original array first.
```suggestion
    const minArrInv = minArr.concat();
    minArrInv.reverse().unshift(x);
```